src/inmanta/db/versions/v202503030.py

```diff
@@ -0,0 +1,238 @@
+            FOREIGN KEY (environment, module_name, module_version)
+                REFERENCES public.module(environment, name, version) ON DELETE CASCADE
+        );
+
+        CREATE INDEX modules_for_agent_environment_agent_name_index
+        ON public.modules_for_agent (environment, agent_name);
+        CREATE INDEX modules_for_agent_environment_module_name_module_version_index
+        ON public.modules_for_agent (environment, module_name, module_version);
+
+    """
+
+    await connection.execute(schema)
+
+    code_data: dict[str, dict[int, dict[str, set[tuple[str, str, str, set[str]]]]]] = defaultdict(
+        lambda: defaultdict(lambda: defaultdict(set))
+    )
+    model_to_module_version_map: dict[int, dict[str, str]] = defaultdict(dict)
+
+    resource_type_to_module: dict[int, dict[str, set[str]]] = defaultdict(lambda: defaultdict(set))
+
+    async def fetch_code_data(
+        code_data: dict[str, dict[int, dict[str, set[tuple[str, str, str, set[str]]]]]],
+        resource_type_to_module: dict[int, dict[str, set[str]]],
+    ) -> None:
+        fetch_source_refs_query = """
+    SELECT DISTINCT environment, version, source_refs, resource from public.code
+        """
+        result = await connection.fetch(fetch_source_refs_query)
+        for res in result:
+            env: str = str(res["environment"])
+            model_version = res["version"]
+            source_refs = res["source_refs"]
+            resource_type: str = str(res["resource"])
+
+            assert isinstance(model_version, int)
+
+            for file_hash, file_data in source_refs.items():  # type: ignore
+                file_path, python_module_name, requirements = file_data
+                inmanta_module_name = python_module_name.split(".")[1]
+                assert isinstance(inmanta_module_name, str)
+                code_data[env][model_version][inmanta_module_name].add(
+                    tuple((file_hash, file_path, python_module_name, frozenset(requirements)))
+                )
+
+                resource_type_to_module[model_version][resource_type].add(inmanta_module_name)
+
+    def build_module_data(
+        code_data: dict[str, dict[int, dict[str, set[tuple[str, str, str, set[str]]]]]],
+        module_data: list[tuple[str, str, str, list[str]]],
+        files_in_module_data: list[tuple[str, str, str, str, str]],
+        model_to_module_version_map: dict[int, dict[str, str]],
+    ) -> None:
+
+        def compute_version_requirements(
+            source_info: set[tuple[str, str, str, set[str]]],
+        ) -> tuple[str, list[str]]:
+            """
+            Helper method
+            """
+            reqs = set()
+            module_version_hash = hashlib.new("sha1")
+
+            for file_hash, _, _, requirements in sorted(source_info, key=lambda x: x[0]):
+                # sort by hash to compute inmanta module version
+                module_version_hash.update(file_hash.encode())
```

Comment:
same files, different requirements will produce the same hash, that is a problem I think

Reply:
Updated the version computation to include the requirements.